### PR TITLE
Adding a counter trace macro to the tracing for getting insights from counters in chip and implemented the macro for ESP32 platform...

### PR DESCRIPTION
**Pull request by shripad621git**
**Change Overview:**

- Added a counter macro to the tracing and all the implementations.
- Implemented the counters for ESP32 platform for deriving insights from them.
- Adding the counters in the chip code at appropriate places to derive insights.

**Testing**
Tested the ESP32 lighting app with and without tracing enabled and checked for the behaviour.

### CODE REVIEW

File: src/tracing/backend.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ class Backend : public ::chip::IntrusiveListNodeBase<>
     /// Trace a zero-sized event
     virtual void TraceInstant(const char * label, const char * group) {}
 
+    virtual void TraceCounter(const char * label, const char * group) {}
```
Comment (andy31415) [inline src/tracing/backend.h:66]: This needs some explanation why a "counter" is needed. If we assume all trace events are captured, a counter seems to be very similar to an instant or anything else that you can count the number of invocations.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Offline discussion: counters are useful, however I am asking for this to not become a "only one backend ever supports this".

If we are to add counter support, other backends should too. In particular perfetto has counters (I see `TRACE_COUNTER` exists) and having logging have something probably also works (e.g. it could be a static atomic variable that gets logged on every call).

File: src/tracing/perfetto/perfetto_tracing.cpp [hunk 2]
```diff
@@ -52,8 +52,28 @@ void PerfettoBackend::LogMessageReceived(MessageReceivedInfo & info)
...
+
+    TRACE_EVENT("Matter", "Counter",         //
+                "Label", label,              //
+                "count", mCounters[counterId] //
+    );
+}
+
 void PerfettoBackend::LogMessageSend(MessageSendInfo & info)
 {
+    ChipLogProgress(SecureChannel, "In trace counter");
```
Comment (andy31415) [inline src/tracing/perfetto/perfetto_tracing.cpp:76]: This debug code should be removed before PR merge

File: src/tracing/perfetto/perfetto_tracing.h [hunk 3]
```diff
@@ -38,12 +40,16 @@ class PerfettoBackend : public ::chip::Tracing::Backend
...
+    void TraceCounter(const char * label, const char * group) override;
     void LogMessageSend(MessageSendInfo &) override;
     void LogMessageReceived(MessageReceivedInfo &) override;
 
     void LogNodeLookup(NodeLookupInfo &) override;
     void LogNodeDiscovered(NodeDiscoveredInfo &) override;
     void LogNodeDiscoveryFailed(NodeDiscoveryFailedInfo &) override;
+
+private:
+    std::unordered_map<std::string, int> mCounters;
```
Comment (andy31415) [inline src/tracing/perfetto/perfetto_tracing.h:52]: perfetto has its own counter macros: https://perfetto.dev/docs/instrumentation/track-events#counters

we should not re-invent them.

Comment (andy31415) [reply src/tracing/perfetto/perfetto_tracing.h:52]: The logic in perfetto is that it does not allow multiplexing, so events, scopes and counters will go via macros_impl only - see comment above. We expect these to be set exclusively.

This is because perfetto did a lot of work on performance tuning so we do not want to bypass that.

File: src/tracing/esp32_trace/counter.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <tracing/esp32_trace/counter.h>
+
+namespace Insights {
+ESPInsightsCounter * ESPInsightsCounter::mHead = nullptr;
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.cpp:22]: some documentation about memory management may be useful here. Specificaly it seems that there is no way to free these, so I assume that is intentional because esp32 does not need it.

File: src/tracing/esp32_trace/counter.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            current->instanceCount++;
+            return current;
+        }
+        previous = current;
+        current  = current->mNext;
+    }
+
+    ESPInsightsCounter * newInstance = new ESPInsightsCounter(label, group);
+    newInstance->mNext               = mHead;
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.cpp:42]: this is unbounded allocation without freeing. Should we at least VerifyOrDie that newInstance is not null?

File: src/tracing/esp32_trace/counter.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#include <esp_log.h>
+#include <string.h>
+
+#define PATH1 "sys.cnt"
+namespace Insights {
+class ESPInsightsCounter
+{
+private:
+    static ESPInsightsCounter * mHead;
+    char label[50];
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:29]: do these need to be static sized? trace macros are generally passed const char *, so these could retain the pointer instead (or do we expect dynamic names/groups?)

File: src/tracing/esp32_trace/counter.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    char group[50];
+    int instanceCount;
+    ESPInsightsCounter * mNext;
+    bool registered = false;
+
+    ESPInsightsCounter(const char * labelParam, const char * groupParam) : instanceCount(1), mNext(nullptr)
+    {
+        strncpy(label, labelParam, sizeof(label));
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:37]: use `chip::Platform::CopyString` to ensure null termination

File: src/tracing/esp32_trace/counter.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+public:
+    static ESPInsightsCounter * GetInstance(const char * label, const char * group);
+
+    int GetInstanceCount() const;
+
+    void ReportMetrics()
+    {
+        if (!registered)
+        {
+            esp_diag_metrics_register("SYS_CNT", label, label, PATH1, ESP_DIAG_DATA_TYPE_UINT);
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:50]: label is repeated here ... should the registration contain group in some way?

File: src/tracing/perfetto/BUILD.gn [hunk 9]
```diff
@@ -75,7 +75,6 @@ static_library("perfetto") {
   public_deps = [
     ":perfetto_tracing",
     "${chip_root}/src/lib/address_resolve",
-    "${chip_root}/src/tracing",
     "${chip_root}/src/transport",
```
Comment (andy31415) [inline src/tracing/perfetto/BUILD.gn:78]: why was this removed? I see `src/tracing/perfetto/perfetto_tracing.h` having an include of `#include <tracing/backend.h>` so I assume the dependencies are needed.

File: src/tracing/esp32_trace/counter.h (hunk 6 above)
Comment (shripad621git) [reply src/tracing/esp32_trace/counter.h:29]: We expect dynamic names for label and group to be assigned based on the trace macro.

File: src/tracing/perfetto/include/matter/tracing/macros_impl.h [hunk 10]
```diff
@@ -30,3 +30,10 @@ PERFETTO_DEFINE_CATEGORIES(perfetto::Category("Matter").SetDescription("Matter t
 #define MATTER_TRACE_END(label, group) TRACE_EVENT_END("Matter")
 #define MATTER_TRACE_INSTANT(label, group) TRACE_EVENT_INSTANT("Matter", label, "class_name", group)
 #define MATTER_TRACE_SCOPE(label, group) TRACE_EVENT("Matter", label, "class_name", group)
+
+#define MATTER_TRACE_COUNTER(label, group)                                                                                         \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        static int count##_label = 0;                                                                                              \
+        TRACE_COUNTER("Matter", perfetto::CounterTrack(label), ++count##_label);                                                   \
```
[the remark this answers is not in the document]
Comment (shripad621git) [reply src/tracing/perfetto/include/matter/tracing/macros_impl.h:38]: Although this also does not have a group, the group has a significant role and may play a significant role in analysis of traces from a particular section , also for filtering out the traces of concern.

File: src/tracing/perfetto/include/matter/tracing/macros_impl.h [hunk 11]
```diff
@@ -30,3 +30,10 @@ PERFETTO_DEFINE_CATEGORIES(perfetto::Category("Matter").SetDescription("Matter t
 #define MATTER_TRACE_END(label, group) TRACE_EVENT_END("Matter")
 #define MATTER_TRACE_INSTANT(label, group) TRACE_EVENT_INSTANT("Matter", label, "class_name", group)
 #define MATTER_TRACE_SCOPE(label, group) TRACE_EVENT("Matter", label, "class_name", group)
+
+#define MATTER_TRACE_COUNTER(label, group)                                                                                         \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        static int count##_label##_group = 0;                                                                                      \
+        TRACE_COUNTER("Matter", perfetto::CounterTrack(label), ++count##_label##_group);                                           \
```
Comment (andy31415) [inline src/tracing/perfetto/include/matter/tracing/macros_impl.h:38]: do we need CounterTrack? Docs at https://android.googlesource.com/platform/external/perfetto/+/refs/heads/master/include/perfetto/tracing/track_event.h#383 seem to say that `TRACE_COUNTER("Matter", label, ...)` would be sufficient.

**Review by andy31415 — APPROVED** [review]

File: src/tracing/esp32_trace/counter.h (hunk 8 above)
Comment (andy31415) [reply src/tracing/esp32_trace/counter.h]: Not resolved:
   - ESP32 implementation makes no use of the group, perfetto makes no use of the group. This has a chance to create non-unique counters. either make use of the group for counters or remove the group alltogether.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Changes requested after checking if previous comments were resolved or not:

- specifically "group" is part of the macro invocation however it is not used for counter naming and as a result it can create overlapping counters (e.g. `MATTER_TRACE_COUNTER("ABC", "A");` and `MATTER_TRACE_COUNTER("ABC", "B");` would create the same perfetto counter and would probably report the same in esp. This is not ok. Either group should be used, or, more likely, group argument should be dropped.

File: src/tracing/esp32_trace/counter.h [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+#include <lib/support/CHIPMemString.h>
+
+#define PATH "insights.cnt"
+
+namespace Insights {
+class ESPInsightsCounter
+{
+private:
+    static ESPInsightsCounter * mHead; // head of the counter list
+    char label[50];                    // unique key
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:31]: my previous comment was lost due to force push.
Re-adding here to be up to date: why is this expected to be dynamic and in what way? I only see static strings so far. Could you give examples?

Asking because for other tracing mechanisms we explicitly require static strings because that allows tracing to be less intrusive.

File: src/tracing/perfetto/include/matter/tracing/macros_impl.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply src/tracing/perfetto/include/matter/tracing/macros_impl.h]: You are creating overlapping counters ... in this case this is actually wrong: two seemingly different TRACE_COUNTER (because they have different group) will create identical TRACE_COUNTER invocations.

I am unclear what you mean by the statement `the group has a significant role and may play a significant role in analysis of traces from a particular section`. Please give an example. If it is important, it should be contained in the macro invocation to make the macro unique. Otherwise you are intentionally making that different groups generate same macro if the label are the same and I do not understand why that is a desirable behavior.

File: src/tracing/esp32_trace/counter.h (hunk 6 above)
Comment (andy31415) [reply src/tracing/esp32_trace/counter.h]: Please provide an example. otherwise please apply `You ain't gonna need it` - if we do not need it now, please do not implement "for the future" because most often the future never needs it. If we do need it, switching from `const char *` to alocated memory and string copy should be easy.

File: src/tracing/esp32_trace/counter.h (hunk 12 above)
Comment (shripad621git) [reply src/tracing/esp32_trace/counter.h]: I have addressed this in my latest commit.

File: src/tracing/perfetto/include/matter/tracing/macros_impl.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (shripad621git) [reply src/tracing/perfetto/include/matter/tracing/macros_impl.h]: I have removed the group from the counter macro.

File: src/tracing/esp32_trace/counter.h (hunk 8 above)
Comment (shripad621git) [reply src/tracing/esp32_trace/counter.h]: Addressed in latest commit.

File: src/tracing/esp32_trace/counter.h (hunk 6 above)
Comment (shripad621git) [reply src/tracing/esp32_trace/counter.h]: I have addressed the comment in the latest commit.

File: src/tracing/esp32_trace/counter.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+ *    limitations under the License.
+ */
+
+#include <esp_diagnostics_metrics.h>
+#include <esp_log.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CHIPMemString.h>
+#include <string.h>
+
+#define PATH "insights.cnt"
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:25]: this define does not seem needed in a header (for others to see). Please it in the cpp file.

Also `PATH` is a very generic name. Make it more specific - is this a file path? is this for esp_insights specifically?

File: src/tracing/esp32_trace/counter.h [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace Insights {
+class ESPInsightsCounter
```
Comment (andy31415) [inline src/tracing/esp32_trace/counter.h:28]: add class documentation: what does this do and how is it used?

File: src/tracing/esp32_trace/counter.h (hunk 13 above)
Comment (andy31415) [reply src/tracing/esp32_trace/counter.h:25]: I am actually unclear on where this is used at all...drop it?

**Review by andy31415 — APPROVED** [review]
